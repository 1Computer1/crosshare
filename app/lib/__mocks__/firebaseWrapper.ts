const firebaseTesting = jest.requireActual('@firebase/testing');
import type firebaseAdminType from 'firebase-admin';

export let App: firebase.app.App;
export let AdminApp: firebaseAdminType.app.App;

export const setApp = (app: firebase.app.App) => { App = app; };
export const setAdminApp = (app: firebaseAdminType.app.App) => { AdminApp = app; };

export const AuthProvider = null;

export const DeleteSentinal = firebaseTesting.firestore.FieldValue.delete();

export const ServerTimestamp = firebaseTesting.firestore.FieldValue.serverTimestamp();

export const TimestampClass = firebaseTesting.firestore.Timestamp;
export const AdminTimestamp = firebaseTesting.firestore.Timestamp;

let anonApp: firebase.app.App;
let anonUser: firebase.User;
export const setUpForSignInAnonymously = (app: firebase.app.App, user: firebase.User) => {
  anonApp = app;
  anonUser = user;
};

export const signInAnonymously = async () => {
  setApp(anonApp);
  return anonUser;
};
