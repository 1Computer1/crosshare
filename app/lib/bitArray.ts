
const b64ToI: Array<number> = [];
const iToB64: Array<string> = [];
const numBase = '0'.charCodeAt(0);
const lowerCaseBase = 'a'.charCodeAt(0);
const upperCaseBase = 'A'.charCodeAt(0);
for (let i = 0; i <= 9; ++i) {
  b64ToI[numBase + i] = i;
  iToB64[i] = String.fromCharCode(numBase + i);
}
for (let i = 0; i < 26; ++i) {
  b64ToI[lowerCaseBase + i] = i + 10;
  iToB64[i + 10] = String.fromCharCode(lowerCaseBase + i);
  b64ToI[upperCaseBase + i] = i + 36;
  iToB64[i + 36] = String.fromCharCode(upperCaseBase + i);
}
b64ToI['-'.charCodeAt(0)] = 62;
iToB64[62] = '-';
b64ToI['_'.charCodeAt(0)] = 63;
iToB64[63] = '_';

function popCount(v: number): number {
  v -= ((v >>> 1) & 0x55555555);
  v = (v & 0x33333333) + ((v >>> 2) & 0x33333333);
  return (((v + (v >>> 4) & 0xF0F0F0F) * 0x1010101) >>> 24);
}

export class BitArray {
  constructor(private nums: Array<number>, private usedInts: number) { }

  static zero() {
    return new this([0], 1);
  }

  static fromString(input: string, base: 32 | 64) {
    if (base === 32) {
      input = input.toLowerCase();
    }
    // We only use 30 bits out of 32 max but it keeps everything simple
    const bitsPerChar = base === 32 ? 5 : 6;
    const charsPerNum = base === 32 ? 6 : 5;

    const nums: Array<number> = [];
    let usedInts = 0;

    for (let i = 0; i < input.length; ++i) {
      const index = input.length - i - 1;
      const x = b64ToI[input.charCodeAt(index)];
      const mod = i % charsPerNum;
      if (mod === 0) {
        nums[usedInts++] = x;
      } else {
        nums[usedInts - 1] |= (x << mod * bitsPerChar);
      }
    }
    return new this(nums, usedInts);
  }

  isZero() {
    return this.usedInts === 0 || (this.usedInts === 1 && this.nums[0] === 0);
  }

  clamp() {
    while (this.usedInts > 0 && this.nums[this.usedInts - 1] === 0) {
      --this.usedInts;
    }
  }

  setBit(index: number) {
    const numIndex = Math.floor(index / 30);
    const rem = index % 30;
    for (let i = this.usedInts; i < numIndex + 1; i += 1) {
      this.usedInts += 1;
      this.nums.push(0);
    }
    this.nums[numIndex] |= (1 << rem);
  }

  toString(base: 32 | 64) {
    const bitsPerChar = base === 32 ? 5 : 6;
    const charsPerNum = base === 32 ? 6 : 5;
    const mask = (1 << bitsPerChar) - 1;
    let s = '';
    for (let i = this.usedInts - 1; i >= 0; --i) {
      for (let j = charsPerNum - 1; j >= 0; --j) {
        const char = iToB64[(this.nums[i] >> (j * bitsPerChar)) & mask];
        if (s || char !== '0') {
          s += char;
        }
      }
    }
    return s || '0';
  }

  bitLength(): number {
    if (!this.usedInts) {
      return 0;
    }
    return 30 * (this.usedInts - 1) + (32 - Math.clz32(this.nums[this.usedInts - 1]));
  }

  /** TODO we can make a faster version if we expect bit array to be sparse */
  bitCount(): number {
    let sum = 0;
    for (let i = 0; i < this.usedInts; i += 1) {
      if (this.nums[i]) {
        sum += popCount(this.nums[i]);
      }
    }
    return sum;
  }

  and(other: BitArray) {
    const nums: Array<number> = [];
    const usedInts = Math.min(this.usedInts, other.usedInts);

    for (let i = 0; i < usedInts; ++i) {
      nums[i] = this.nums[i] & other.nums[i];
    }

    const res = new BitArray(nums, usedInts);
    res.clamp();
    return res;
  }

  activeBits(): Array<number> {
    const ret = [];
    for (let i = this.usedInts - 1; i >= 0; i--) {
      let num = this.nums[i];
      while (num !== 0) {
        const t = 31 - Math.clz32(num);
        num ^= 1 << t;
        ret.push((i * 30) + t);
      }
    }
    return ret;
  }
}
