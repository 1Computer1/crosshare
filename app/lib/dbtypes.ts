import * as t from 'io-ts';

// We import this instead of defining it here so that we can define it
// differently for the main react app and the 'functions' subdirectory
import { timestamp } from './timestamp';
export { timestamp } from './timestamp';

const CommentV = t.type({
  /** comment text */
  c: t.string,
  /** author id */
  a: t.string,
  /** author display name */
  n: t.string,
  /** author solve time in fractional seconds */
  t: t.number,
  /** author did cheat? */
  ch: t.boolean,
  /** comment publish timestamp */
  p: timestamp
});
type CommentT = t.TypeOf<typeof CommentV>;

export interface CommentWithRepliesT extends CommentT {
  /** comment id */
  i: string,
  /** replies */
  r?: Array<CommentWithRepliesT>
}

const CommentWithRepliesV: t.Type<CommentWithRepliesT> = t.recursion('CommentWithReplies', () =>
  t.intersection([
    CommentV,
    t.type({
      /** comment id */
      i: t.string,
    }),
    t.partial({
      /** replies */
      r: t.array(CommentWithRepliesV)
    })
  ])
);

export const CommentForModerationV = t.intersection([
  CommentV,
  t.type({
    /** puzzle id */
    pid: t.string,
    /** id of the comment this is a reply to */
    rt: t.union([t.string, t.null])
  })
]);
export type CommentForModerationT = t.TypeOf<typeof CommentForModerationV>;

export const CommentForModerationWithIdV = t.intersection([
  CommentForModerationV,
  t.type({
    /** comment id */
    i: t.string,
  })
]);
export type CommentForModerationWithIdT = t.TypeOf<typeof CommentForModerationWithIdV>;

const DBPuzzleMandatoryV = t.type({
  /** created at */
  ca: timestamp,
  /** author's user id */
  a: t.string,
  /** author's display name */
  n: t.string,
  /** category */
  c: t.union([t.string, t.null]),
  /** is this puzzle moderated? */
  m: t.boolean,
  /** timestamp when the puzzle goes live */
  p: t.union([timestamp, t.null]),
  /** title */
  t: t.string,
  /** grid width / columns */
  w: t.number,
  /** grid height / rows */
  h: t.number,
  /** across clue strings */
  ac: t.array(t.string),
  /** across clue display numbers */
  an: t.array(t.number),
  /** down clue strings */
  dc: t.array(t.string),
  /** down clue display numbers */
  dn: t.array(t.number),
  /** grid (solution) */
  g: t.array(t.string),
});
const DBPuzzleOptionalV = t.partial({
  /** highlighted cell indexes */
  hs: t.array(t.number),
  /** use shade instead of circle for highlight? */
  s: t.boolean,
  /** comments */
  cs: t.array(CommentWithRepliesV),
  /** is this puzzle featured (should appear on homepage) */
  f: t.boolean,
});
export const DBPuzzleV = t.intersection([DBPuzzleMandatoryV, DBPuzzleOptionalV]);
export type DBPuzzleT = t.TypeOf<typeof DBPuzzleV>;

export const TimestampedPuzzleV = downloadTimestamped(DBPuzzleV);
export type TimestampedPuzzleT = t.TypeOf<typeof TimestampedPuzzleV>;

const PlayBaseV = t.type({
  /** crossword id */
  c: t.string,
  /** updated at */
  ua: timestamp,
  /** filled in grid */
  g: t.array(t.string),
  /** play time of last update to each cell, in fractional seconds */
  ct: t.array(t.number),
  /** update iteration count per cell */
  uc: t.array(t.number),
  /** list of (checked) correct cells */
  vc: t.array(t.number),
  /** list of (checked) wrong cells */
  wc: t.array(t.number),
  /** list of cells ever marked wrong */
  we: t.array(t.number),
  /** list of revealed cells */
  rc: t.array(t.number),
  /** total play time, in fractional seconds */
  t: t.number,
  /** did "cheat"? */
  ch: t.boolean,
  /** finished the puzzle? */
  f: t.boolean,
});

export const LegacyPlayV = t.intersection([
  PlayBaseV,
  t.type({
    /** user id */
    u: t.string,
  }),
  t.partial({
    /** puzzle title, optional for legacy plays in the db */
    n: t.string,
  })
]);
export type LegacyPlayT = t.TypeOf<typeof LegacyPlayV>;

export const PlayV = t.intersection([
  PlayBaseV,
  t.type({
    /** user id */
    u: t.string,
    /** puzzle title */
    n: t.string,
  })
]);
export type PlayT = t.TypeOf<typeof PlayV>;

// We don't need a user id in local storage, but we do need a title
export const PlayWithoutUserV = t.intersection([
  PlayBaseV,
  t.type({
    /** puzzle title */
    n: t.string,
  })
]);
export type PlayWithoutUserT = t.TypeOf<typeof PlayWithoutUserV>;

export function downloadTimestamped<A>(type: t.Type<A>) {
  return t.type({
    downloadedAt: timestamp,
    data: type,
  });
}

export function downloadOptionallyTimestamped<A>(type: t.Type<A>) {
  return t.type({
    downloadedAt: t.union([timestamp, t.null]),
    data: type,
  });
}

export function getDateString(pd: Date) {
  return pd.getUTCFullYear() + '-' + pd.getUTCMonth() + '-' + pd.getUTCDate();
}

export function prettifyDateString(dateString: string) {
  const groups = dateString.match(/^(\d+)-(\d+)-(\d+)$/);
  if (!groups) {
    throw new Error('Bad date string: ' + dateString);
  }
  return (parseInt(groups[2]) + 1) + '/' + parseInt(groups[3]) + '/' + groups[1];
}

export const PuzzleStatsV = t.type({
  /** author id, denormalized for security rules purposes. */
  a: t.string,
  /** updated at */
  ua: timestamp,
  /** total completions */
  n: t.number,
  /** total completions without cheats */
  s: t.number,
  /** total time spent on puzzle, in fractional seconds */
  nt: t.number,
  /** total time spent by those w/o cheats */
  st: t.number,
  /** total play time of last update to each cell, in fractional seconds */
  ct: t.array(t.number),
  /** total update iteration count for each cell */
  uc: t.array(t.number),
});
export type PuzzleStatsT = t.TypeOf<typeof PuzzleStatsV>;

export const DailyStatsV = t.type({
  /** updated at */
  ua: timestamp,
  /** total completions */
  n: t.number,
  /** user ids with completions */
  u: t.array(t.string),
  /** completions by puzzleId */
  c: t.record(t.string, t.number),
  /** completions by hour (as UTC 0-23) */
  h: t.array(t.number),
});
export type DailyStatsT = t.TypeOf<typeof DailyStatsV>;

export const CronStatusV = t.type({
  ranAt: timestamp,
});
export type CronStatusT = t.TypeOf<typeof CronStatusV>;

/** created at, title */
const AuthoredPuzzleV = t.tuple([timestamp, t.string]);
export type AuthoredPuzzleT = t.TypeOf<typeof AuthoredPuzzleV>;

/** keys are puzzle ids */
export const AuthoredPuzzlesV = t.record(t.string, AuthoredPuzzleV);

/** date string -> puzzle id */
export const CategoryIndexV = t.record(t.string, t.string);
export type CategoryIndexT = t.TypeOf<typeof CategoryIndexV>;
