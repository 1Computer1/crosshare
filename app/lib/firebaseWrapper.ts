declare var firebase: typeof import('firebase');

// On the server we need to require firebase - on client it's done w/ script tags
if (typeof window === "undefined") {
  globalThis.firebase = require("firebase");
  require("firebase/firestore");
}

const firebaseConfig = {
  apiKey: "AIzaSyBrmmBf91peVT5T_Z7N3z9oizsPH5u2pUc",
  authDomain: "auth.crosshare.org",
  databaseURL: "https://mdcrosshare.firebaseio.com",
  projectId: "mdcrosshare",
  storageBucket: "mdcrosshare.appspot.com",
  messagingSenderId: "603173482014",
  appId: "1:603173482014:web:98d7d820731b7c5eaa080f",
  measurementId: "G-LTLN7Z4XBS"
};

// Initialize Firebase
export let App: firebase.app.App;
if (firebase.apps.length) {
  App = firebase.apps[0];
} else {
  App = firebase.initializeApp(firebaseConfig);
  if (typeof window !== "undefined") {
    if (process.env.NODE_ENV === 'production') {
      App.performance();
    }
  }
  if (process.env.NEXT_PUBLIC_USE_FIREBASE_EMULATOR) {
    var db = App.firestore();
    db.settings({
      host: "localhost:8080",
      ssl: false
    });
  }
}

export const setApp = (app: firebase.app.App) => { App = app };

export const AuthProvider = new firebase.auth.GoogleAuthProvider();

export const DeleteSentinal = firebase.firestore.FieldValue.delete();

export const TimestampClass = firebase.firestore.Timestamp;
export type TimestampType = firebase.firestore.Timestamp;
