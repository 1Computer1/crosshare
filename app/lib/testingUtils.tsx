import { ReactElement, ComponentType } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { matchers, createSerializer } from 'jest-emotion';

import { AuthContext } from '../components/AuthContext';

import type firebaseTypes from 'firebase';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);
expect.addSnapshotSerializer(createSerializer());

export const anonymousUser = {
  uid: 'anonymous-user-id',
  isAnonymous: true
} as firebaseTypes.User;

const WithAllProviders: ComponentType = ({ children }) => {
  return (
    <AuthContext.Provider value={{ user: anonymousUser, isAdmin: false, loadingUser: false, error: undefined }}>
      {children}
    </AuthContext.Provider>
  )
}

function wrappedRender(
  ui: ReactElement,
  options?: RenderOptions
): RenderResult {

  return render(ui, {
    wrapper: WithAllProviders,
    ...options,
  });
}
// re-export everything
export * from '@testing-library/react';

// override render method
export { wrappedRender as render };
