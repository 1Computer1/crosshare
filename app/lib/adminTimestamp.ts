/* this is the firebase timestamp type for the main react project! */

import * as t from 'io-ts';
import { either } from 'fp-ts/lib/Either';

import { AdminTimestamp } from './firebaseWrapper';
import type firebaseAdminType from 'firebase-admin';

const isFirestoreTimestamp = (u: unknown): u is firebaseAdminType.firestore.Timestamp =>
  u ? u instanceof AdminTimestamp : false;

const validateTimestamp: t.Validate<unknown, firebaseAdminType.firestore.Timestamp> = (i, c) => {
  if (isFirestoreTimestamp(i)) {
    return t.success(i);
  }
  return either.chain(
    t.type({ seconds: t.number, nanoseconds: t.number }).validate(i, c),
    obj => t.success(new AdminTimestamp(obj.seconds, obj.nanoseconds))
  );
};

export const adminTimestamp = new t.Type<firebaseAdminType.firestore.Timestamp>(
  'AdminTimestamp',
  isFirestoreTimestamp,
  validateTimestamp,
  t.identity,
);
