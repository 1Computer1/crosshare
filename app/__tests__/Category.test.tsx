import React from 'react';
import { anonymousUser, render } from '../lib/testingUtils';
import CategoryPage, { propsForCategoryId } from '../pages/categories/[categoryId]';
import { setApp, TimestampClass } from '../lib/firebaseWrapper';
import * as firebaseTesting from '@firebase/testing';
import { PlayT, CategoryIndexT } from '../lib/dbtypes';

let adminApp: firebase.app.App;
let app: firebase.app.App;
let loggedInApp: firebase.app.App;

jest.mock('../lib/firebaseWrapper');

beforeAll(async () => {
  adminApp = firebaseTesting.initializeAdminApp({ projectId: 'categorytest' }) as firebase.app.App;
  app = firebaseTesting.initializeTestApp({ projectId: 'categorytest' }) as firebase.app.App;
  loggedInApp = firebaseTesting.initializeTestApp({
    projectId: 'categorytest',
    auth: {
      uid: 'anonymous-user-id', admin: false, firebase: {
        sign_in_provider: 'anonymous'
      }
    }
  }) as firebase.app.App;
  setApp(app);

  const dailymini: CategoryIndexT = {
    '2020-4-10': 'foobar',
    '2020-4-1': 'baz',
    '2020-4-5': 'mike'
  };
  await adminApp.firestore().collection('categories').doc('dailymini').set(dailymini);

  const play1: PlayT = {
    c: 'foobar',
    u: 'anonymous-user-id',
    ua: TimestampClass.now(),
    g: [],
    ct: [],
    uc: [],
    vc: [],
    wc: [],
    we: [],
    rc: [],
    t: 70,
    ch: false,
    f: true
  };
  await adminApp.firestore().collection('p').doc('foobar-anonymous-user-id').set(play1);

  const play2: PlayT = {
    c: 'mike',
    u: 'anonymous-user-id',
    ua: TimestampClass.now(),
    g: [],
    ct: [],
    uc: [],
    vc: [],
    wc: [],
    we: [],
    rc: [],
    t: 44,
    ch: false,
    f: false
  };
  await adminApp.firestore().collection('p').doc('mike-anonymous-user-id').set(play2);
});

afterAll(async () => {
  await app.delete();
  await adminApp.delete();
  await loggedInApp.delete();
});

test('basic category page test', async () => {
  const props = await propsForCategoryId('dailymini');
  const { findByText, container } = render(<CategoryPage {...props} />, {});

  const link = await findByText('Daily Mini for 5/10/2020');
  expect(link).toBeInTheDocument();
  expect(link).toHaveAttribute('href', '/crosswords/foobar');
  expect(container).toMatchSnapshot();
});

test('category page when logged in with some plays', async () => {
  setApp(loggedInApp);

  const props = await propsForCategoryId('dailymini');
  const { findByText, container } = render(<CategoryPage {...props} />, { user: anonymousUser });

  expect(await findByText(/completed without helpers/)).toBeInTheDocument();
  expect(await findByText(/unfinished/)).toBeInTheDocument();
  expect(container).toMatchSnapshot();

  setApp(app);
});

test('category 404', async () => {
  const props = await propsForCategoryId('dailyfulllength');
  const { getByText } = render(<CategoryPage {...props} />, {});

  expect(getByText(/invalid category/i)).toBeVisible();
});
