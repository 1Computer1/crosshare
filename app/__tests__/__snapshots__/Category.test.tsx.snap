// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic category page test 1`] = `
.emotion-6 {
  height: 35px;
  background-color: var(--primary);
}

.emotion-5 {
  padding: 0 10px;
  height: 100%;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 31px;
}

.emotion-1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  cursor: pointer;
}

.emotion-0 {
  margin-left: 5px;
  display: none;
  color: var(--text);
  font-size: 25px;
}

@media (min-width:576px) {
  .emotion-0 {
    display: inline-block;
  }
}

.emotion-4 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline;
  margin: 0;
  padding: 0 0.45em;
  color: var(--text);
}

.emotion-4:hover,
.emotion-4:focus {
  color: var(--text);
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: var(--top-bar-hover);
}

.emotion-2 {
  vertical-align: baseline;
  font-size: 25px;
}

.emotion-3 {
  margin-left: 5px;
  vertical-align: middle;
  display: none;
  font-size: 15px;
}

@media (min-width:576px) {
  .emotion-3 {
    display: inline-block;
  }
}

.emotion-16 {
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.emotion-9 {
  padding: 0.5em 0;
  width: 100%;
}

.emotion-7 {
  display: inline-block;
  width: 50%;
  text-align: right;
  padding-right: 1em;
  font-weight: bold;
}

.emotion-8 {
  display: inline-block;
  width: 50%;
  padding-left: 1em;
}

<div>
  <header
    class="emotion-6"
  >
    <div
      class="emotion-5"
    >
      <a
        class="emotion-1"
        href="/"
        title="Crosshare Home"
      >
        <svg
          height="31"
          viewBox="0 0 16 16"
          width="31"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="var(--logo-white)"
          >
            <path
              d="M1 10h14v5H1z"
            />
            <path
              d="M3 9h10v1H3zm0-7h5v3H3z"
            />
            <path
              d="M6 5h1v5H6zm4-2h2v6h-2z"
            />
            <path
              d="M12 1h2v3h-2z"
            />
            <path
              d="M14 1h1v1h-1z"
            />
          </g>
          <path
            d="M13 0h3v1h-3zM4 1h3v1H4z"
          />
          <path
            d="M12 1h1v1h-1zm3 0h1v1h-1zM3 2h1v1H3zm3 0h2v1H6z"
          />
          <path
            d="M11 2h1v1h-1zm3 0h1v1h-1zM2 3h1v3H2zm3 0h1v6H5z"
          />
          <path
            d="M7 3h1v7H7zm3 0h1v1h-1zm3 0h1v1h-1zM4 4h1v1H4z"
          />
          <path
            d="M9 4h1v5H9zm3 0h1v1h-1z"
          />
          <path
            d="M3 5h1v1H3zm8 0h1v4h-1zM4 8h1v1H4z"
          />
          <path
            d="M3 9h1v1H3zm5 0h1v1H8z"
          />
          <path
            d="M12 9h1v1h-1zM1 10h2v1H1z"
          />
          <path
            d="M5 10h1v1H5zm8 0h2v1h-2z"
          />
          <path
            d="M0 11h1v3H0zm15 0h1v3h-1zm-9 0h1v1H6z"
          />
          <path
            d="M4 12h2v1H4z"
          />
          <path
            d="M5 13h2v1H5zm4 0h2v1H9zm-8 1h3v1H1zm11 0h3v1h-3z"
          />
          <path
            d="M4 15h8v1H4z"
          />
          <rect
            height="1"
            width="1"
            x="10"
            y="10"
          >
            <animate
              attributeName="width"
              begin="10;a.end+10"
              calcMode="discrete"
              dur="0.5s"
              id="a"
              values="1;0"
            />
          </rect>
          <rect
            height="1"
            width="1"
            x="9"
            y="11"
          >
            <animate
              attributeName="width"
              begin="10;b.end+10"
              calcMode="discrete"
              dur="0.5s"
              id="b"
              values="1;3"
            />
          </rect>
          <rect
            height="1"
            width="2"
            x="10"
            y="12"
          >
            <animate
              attributeName="width"
              begin="10;c.end+10"
              calcMode="discrete"
              dur="0.5s"
              id="c"
              values="2;0"
            />
          </rect>
        </svg>
        <span
          class="emotion-0"
        >
          CROSSHARE
        </span>
      </a>
      
      <a
        class="emotion-4"
        href="/account"
        title="Account"
      >
        <span
          class="emotion-2"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
        </span>
        <span
          class="emotion-3"
        >
          Account
        </span>
      </a>
    </div>
  </header>
  <ul
    class="emotion-16"
  >
    <li
      class="emotion-9"
    >
      <a
        class="emotion-7"
        href="/crosswords/foobar"
      >
        Daily Mini for 5/10/2020
      </a>
      <div
        class="emotion-8"
      />
    </li>
    <li
      class="emotion-9"
    >
      <a
        class="emotion-7"
        href="/crosswords/mike"
      >
        Daily Mini for 5/5/2020
      </a>
      <div
        class="emotion-8"
      />
    </li>
    <li
      class="emotion-9"
    >
      <a
        class="emotion-7"
        href="/crosswords/baz"
      >
        Daily Mini for 5/1/2020
      </a>
      <div
        class="emotion-8"
      />
    </li>
  </ul>
</div>
`;

exports[`category page when logged in with some plays 1`] = `
.emotion-6 {
  height: 35px;
  background-color: var(--primary);
}

.emotion-5 {
  padding: 0 10px;
  height: 100%;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 31px;
}

.emotion-1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  cursor: pointer;
}

.emotion-0 {
  margin-left: 5px;
  display: none;
  color: var(--text);
  font-size: 25px;
}

@media (min-width:576px) {
  .emotion-0 {
    display: inline-block;
  }
}

.emotion-4 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline;
  margin: 0;
  padding: 0 0.45em;
  color: var(--text);
}

.emotion-4:hover,
.emotion-4:focus {
  color: var(--text);
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: var(--top-bar-hover);
}

.emotion-2 {
  vertical-align: baseline;
  font-size: 25px;
}

.emotion-3 {
  margin-left: 5px;
  vertical-align: middle;
  display: none;
  font-size: 15px;
}

@media (min-width:576px) {
  .emotion-3 {
    display: inline-block;
  }
}

.emotion-16 {
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.emotion-9 {
  padding: 0.5em 0;
  width: 100%;
}

.emotion-10 {
  display: inline-block;
  width: 50%;
  text-align: right;
  padding-right: 1em;
  font-weight: bold;
}

.emotion-8 {
  display: inline-block;
  width: 50%;
  padding-left: 1em;
}

.emotion-7 {
  display: inline-block;
  width: 50%;
  text-align: right;
  padding-right: 1em;
  font-weight: normal;
}

<div>
  <header
    class="emotion-6"
  >
    <div
      class="emotion-5"
    >
      <a
        class="emotion-1"
        href="/"
        title="Crosshare Home"
      >
        <svg
          height="31"
          viewBox="0 0 16 16"
          width="31"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="var(--logo-white)"
          >
            <path
              d="M1 10h14v5H1z"
            />
            <path
              d="M3 9h10v1H3zm0-7h5v3H3z"
            />
            <path
              d="M6 5h1v5H6zm4-2h2v6h-2z"
            />
            <path
              d="M12 1h2v3h-2z"
            />
            <path
              d="M14 1h1v1h-1z"
            />
          </g>
          <path
            d="M13 0h3v1h-3zM4 1h3v1H4z"
          />
          <path
            d="M12 1h1v1h-1zm3 0h1v1h-1zM3 2h1v1H3zm3 0h2v1H6z"
          />
          <path
            d="M11 2h1v1h-1zm3 0h1v1h-1zM2 3h1v3H2zm3 0h1v6H5z"
          />
          <path
            d="M7 3h1v7H7zm3 0h1v1h-1zm3 0h1v1h-1zM4 4h1v1H4z"
          />
          <path
            d="M9 4h1v5H9zm3 0h1v1h-1z"
          />
          <path
            d="M3 5h1v1H3zm8 0h1v4h-1zM4 8h1v1H4z"
          />
          <path
            d="M3 9h1v1H3zm5 0h1v1H8z"
          />
          <path
            d="M12 9h1v1h-1zM1 10h2v1H1z"
          />
          <path
            d="M5 10h1v1H5zm8 0h2v1h-2z"
          />
          <path
            d="M0 11h1v3H0zm15 0h1v3h-1zm-9 0h1v1H6z"
          />
          <path
            d="M4 12h2v1H4z"
          />
          <path
            d="M5 13h2v1H5zm4 0h2v1H9zm-8 1h3v1H1zm11 0h3v1h-3z"
          />
          <path
            d="M4 15h8v1H4z"
          />
          <rect
            height="1"
            width="1"
            x="10"
            y="10"
          >
            <animate
              attributeName="width"
              begin="10;a.end+10"
              calcMode="discrete"
              dur="0.5s"
              id="a"
              values="1;0"
            />
          </rect>
          <rect
            height="1"
            width="1"
            x="9"
            y="11"
          >
            <animate
              attributeName="width"
              begin="10;b.end+10"
              calcMode="discrete"
              dur="0.5s"
              id="b"
              values="1;3"
            />
          </rect>
          <rect
            height="1"
            width="2"
            x="10"
            y="12"
          >
            <animate
              attributeName="width"
              begin="10;c.end+10"
              calcMode="discrete"
              dur="0.5s"
              id="c"
              values="2;0"
            />
          </rect>
        </svg>
        <span
          class="emotion-0"
        >
          CROSSHARE
        </span>
      </a>
      
      <a
        class="emotion-4"
        href="/account"
        title="Account"
      >
        <span
          class="emotion-2"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
        </span>
        <span
          class="emotion-3"
        >
          Account
        </span>
      </a>
    </div>
  </header>
  <ul
    class="emotion-16"
  >
    <li
      class="emotion-9"
    >
      <a
        class="emotion-7"
        href="/crosswords/foobar"
      >
        Daily Mini for 5/10/2020
      </a>
      <div
        class="emotion-8"
      >
        completed without helpers
      </div>
    </li>
    <li
      class="emotion-9"
    >
      <a
        class="emotion-10"
        href="/crosswords/mike"
      >
        Daily Mini for 5/5/2020
      </a>
      <div
        class="emotion-8"
      >
        unfinished
      </div>
    </li>
    <li
      class="emotion-9"
    >
      <a
        class="emotion-10"
        href="/crosswords/baz"
      >
        Daily Mini for 5/1/2020
      </a>
      <div
        class="emotion-8"
      />
    </li>
  </ul>
</div>
`;
